File: project-pgkram/frontend/src/api/analytics.js
import { apiRequest } from '../utils/api';

const BASE_URL = import.meta.env.VITE_BACKEND_URL || 'http://localhost:8000';

export async function fetchJobStats(signal) {
  try {
    return await apiRequest(`${BASE_URL}/api/jobs/stats`, { signal }, 15000);
  } catch (error) {
    if (error.name === 'AbortError') {
      throw error; // Re-throw AbortError as-is
    }
    throw new Error(`Failed to fetch job stats: ${error.message}`);
  }
}

export async function searchJobs(payload, signal) {
  try {
    return await apiRequest(`${BASE_URL}/api/jobs/search`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
      signal,
    }, 15000);
  } catch (error) {
    if (error.name === 'AbortError') {
      throw error; // Re-throw AbortError as-is
    }
    throw new Error(`Failed to search jobs: ${error.message}`);
  }
}

export async function fetchTrends(months = 6, signal) {
  try {
    return await apiRequest(`${BASE_URL}/api/analytics/trends?months=${months}`, { signal }, 15000);
  } catch (error) {
    if (error.name === 'AbortError') {
      throw error; // Re-throw AbortError as-is
    }
    throw new Error(`Failed to fetch trends: ${error.message}`);
  }
}

export async function fetchTopDistricts(limit = 5, signal) {
  try {
    return await apiRequest(`${BASE_URL}/api/analytics/top-districts?limit=${limit}`, { signal }, 15000);
  } catch (error) {
    if (error.name === 'AbortError') {
      throw error; // Re-throw AbortError as-is
    }
    throw new Error(`Failed to fetch top districts: ${error.message}`);
  }
}

export async function fetchSalaryDistribution(signal) {
  try {
    return await apiRequest(`${BASE_URL}/api/analytics/salary-distribution`, { signal }, 15000);
  } catch (error) {
    if (error.name === 'AbortError') {
      throw error; // Re-throw AbortError as-is
    }
    throw new Error(`Failed to fetch salary distribution: ${error.message}`);
  }
}

export async function fetchTopOrganizations(limit = 5, signal) {
  try {
    return await apiRequest(`${BASE_URL}/api/analytics/top-organizations?limit=${limit}`, { signal }, 15000);
  } catch (error) {
    if (error.name === 'AbortError') {
      throw error; // Re-throw AbortError as-is
    }
    throw new Error(`Failed to fetch top organizations: ${error.message}`);
  }
}

File: project-pgkram/backend/routers/analytics.py
from fastapi import APIRouter, HTTPException
from datetime import datetime

from database import db_manager
from models import JobSearchRequest, JobSearchResponse


router = APIRouter(prefix="/api", tags=["analytics"])


@router.get("/jobs/stats")
async def get_job_statistics():
    """Get job portal statistics"""
    try:
        stats = await db_manager.get_job_statistics()
        return {
            "success": True,
            "data": stats,
            "timestamp": datetime.now().isoformat()
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching statistics: {str(e)}")


@router.post("/jobs/search", response_model=JobSearchResponse)
async def search_jobs(request: JobSearchRequest):
    """Search jobs based on criteria"""
    try:
        jobs = await db_manager.search_jobs(
            job_type=request.job_type,
            qualification=request.qualification,
            district=request.district,
            experience=request.experience,
            limit=request.limit
        )

        return JobSearchResponse(
            success=True,
            data=jobs,
            count=len(jobs),
            message=f"Found {len(jobs)} jobs matching your criteria"
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error searching jobs: {str(e)}")


@router.get("/analytics/trends")
async def trends(months: int = 6):
    try:
        data = await db_manager.get_jobs_trend_by_month(months)
        return {"success": True, "data": data}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching trends: {str(e)}")


@router.get("/analytics/top-districts")
async def top_districts(limit: int = 5):
    try:
        data = await db_manager.get_top_districts(limit)
        return {"success": True, "data": data}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching top districts: {str(e)}")


@router.get("/analytics/salary-distribution")
async def salary_distribution():
    try:
        data = await db_manager.get_salary_distribution()
        return {"success": True, "data": data}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching salary distribution: {str(e)}")


@router.get("/analytics/top-organizations")
async def top_organizations(limit: int = 5):
    try:
        data = await db_manager.get_top_organizations(limit)
        return {"success": True, "data": data}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching organizations: {str(e)}")


